core/services/ocr2/plugins/ocr2keeper/evm21/feed_lookup.go

```diff
@@ -0,0 +1,442 @@
+
+		// if it doesn't decode to the FeedLookup custom error continue/skip
+		feedLookup, err := r.decodeFeedLookup(upkeepResults[i].PerformData)
+		if err != nil {
+			r.lggr.Errorf("[FeedLookup] upkeep %s block %d decodeFeedLookup: %v", upkeepId, block, err)
+			continue
+		}
+		r.lggr.Infof("[FeedLookup] upkeep %s block %d feedLookup=%v", upkeepId, block, feedLookup)
+
+		// do the mercury lookup request
+		values, retryable, err := r.doMercuryRequest(ctx, feedLookup, upkeepId)
+		if err != nil {
+			r.lggr.Errorf("[FeedLookup] upkeep %s block %d doMercuryRequest: %v", upkeepId, block, err)
+			r.lggr.Infof("[FeedLookup] upkeep %s block %d retryable: %s", upkeepId, block, retryable)
+			upkeepResults[i].Retryable = retryable
+			continue
+		}
+
+		r.lggr.Debugf("[FeedLookup] upkeep %s block %d values: %v", values)
+		r.lggr.Debugf("[FeedLookup] upkeep %s block %d extraData: %v", feedLookup.extraData)
+		mercuryBytes, err := r.checkCallback(ctx, upkeepId, values, feedLookup.extraData, block)
+		if err != nil {
+			r.lggr.Errorf("[FeedLookup] upkeep %s block %d checkCallback err: %v", upkeepId, block, err)
+			continue
+		}
+
+		r.lggr.Infof("FeedLookup checkCallback b=%v", mercuryBytes.String())
+		r.lggr.Infof("FeedLookup checkCallback b=%v", mercuryBytes)
+		needed, performData, failureReason, gasUsed, err := r.packer.UnpackCheckCallbackResult(mercuryBytes)
+		if err != nil {
+			r.lggr.Errorf("[FeedLookup] upkeep %s block %d UnpackCheckCallbackResult err: %v", upkeepId, block, err)
+		}
+		r.lggr.Debugf("[FeedLookup] upkeep %s block %d needed %v\nperformData: %v\nfailureReason: %d\ngasUsed: %s\nperformData: %s", upkeepId, block, needed, performData, failureReason, gasUsed.String(), hexutil.Encode(performData))
+
+		if int(failureReason) == UPKEEP_FAILURE_REASON_MERCURY_CALLBACK_REVERTED {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_MERCURY_CALLBACK_REVERTED
+			r.lggr.Debugf("[FeedLookup] upkeep %s block %d mercury callback reverts", upkeepId, block)
+			continue
+		}
+
+		if !needed {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_UPKEEP_NOT_NEEDED
+			r.lggr.Debugf("[FeedLookup] upkeep %s block %d callback reports upkeep not needed", upkeepId, block)
+			continue
+		}
+
+		upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_NONE
+		upkeepResults[i].Eligible = true
+		upkeepResults[i].PerformData = performData
+		r.lggr.Infof("[FeedLookup] upkeep %s block %d successful with perform data: %+v", upkeepId, block, performData)
+	}
+	// don't surface error to plugin bc FeedLookup process should be self-contained.
+	return upkeepResults, nil
+}
+
+// allowedToUseMercury retrieves upkeep's administrative offchain config and decode a mercuryEnabled bool to indicate if
+// this upkeep is allowed to use Mercury service.
+func (r *EvmRegistry) allowedToUseMercury(opts *bind.CallOpts, upkeepId *big.Int) (bool, error) {
+	allowed, ok := r.mercury.allowListCache.Get(upkeepId.String())
```

Comment:
cc @amirylm if this can be combined with the upkeep life cycle management component used for log filters

Reply:
Yep, for all information we want to fetch from the contract w.r.t upkeeps (`AdminOffchainConfig`, `TriggerConfig`...)
Note I'll do it after we have everything in place for the happy path, so for now we can keep this as is.